### DISCUSSION TITLE
using the errors "type" field

### DISCUSSION THREAD

**Discussion by sezanzeb** [pydantic/pydantic#5112] [Question]
The docs say

> type
> a computer-readable identifier of the error type.

(https://docs.pydantic.dev/usage/models/#error-handling)

Unfortunately type is all lowecase instead of using the actual error class name, and it also prepends the base class name.

For example, if I throw this error:

```py
class ErrorFoo(ValueError):
    pass
```

type will be `value_error.errorfoo`

How can I programmatically compare an error class with the error type? What about this string is computer-readable?

Why is the "type" field not containing the error class instead? Like `{"type": ErrorFoo}`

I want to do something like

```py
if error["type"] == ErrorFoo:
    stuff()
```

to do this, I would currently have to

1. get the base class of ErrorFoo
2. transform it to snake_case
3. lowercase the error class name
4. concatenate the two

To be honest, unless I'm missing something this is not at all computer-readable. `"ValueError.ErrorFoo"` would have been a lot better.

**Comment by samuelcolvin** [comment]
The intention was not to make it easy to compare the error type to exceptions, but rather to perform specific actions based on type. Or use type in some kind of lookup.

We aren't claiming "computer-readable" means "ready to compare to the exception class.

To do what you want, it would probably be easier to have a lookup dict with type as key and exception as value.

**Reply by sezanzeb** [reply]
> To do what you want, it would probably be easier to have a lookup dict with type as key and exception as value.

```
{
    'value_error.errorfoo': ErrorFoo
}
```

No. That means as soon as someone refactors `ErrorFoo` to be `ErrorBar`, without knowing about this dict, the code will break.

**Reply by sezanzeb** [reply]
I'm using `ErrorFoo.__name__.lower() in error["type"]` now.

**Reply by sezanzeb** [reply]
> We aren't claiming "computer-readable" means "ready to compare to the exception class.

A computer can't read things if they are unrecognizable. This is just one example, there could be other use-cases.

> The intention was not to make it easy to compare the error type to exceptions, but rather to perform specific actions based on type. Or use type in some kind of lookup.

Lookup involves comparing.

Unless you hardcode the lookup table (bad practice, as explained above) there is no way to do a lookup with this.

**Reply by sezanzeb** [reply]
Another problem: "FooError" is stripped of "Error". This makes it even harder to work with it. So all I am left with is "foo"

**Comment by samuelcolvin** [comment]
I'm on my phone, but you should be about to find in our source where we do the conversion, and reverse it.

**Reply by sezanzeb** [reply]
Thanks for the suggestion. But I'm afraid reversing "foobar" to "fooBarError" is impossible. I can't guess where the camelCase was.

I'd rather have to transform my own class name to the pydantic-error-type-name. I'm doing something like this right now, it works sufficiently well.

```py
def pydantify(error: type):
    """Generate a string as it would appear IN pydantic error types.

    This does not include the base class name, which is transformed to snake case in
    pydantic. Example pydantic error type: "value_error.foobar" for FooBarError.
    """
    lower_classname = error.__name__.lower()
    if lower_classname.endswith('error'):
        return lower_classname[:-len('error')]
    return lower_classname

if pydantify(FooBarError) in error["type"]:
    ...
```

Or something like `def isErrorType(pydantic_error: dict, error: type) -> bool` would be thinkable as well.

**Reply by samuelcolvin** [reply]
Makes sense, this is what I would have proposed.

How errors work is changing completely in v2, we can probably keep the entire exception name in ctx is that would help.

**Reply by sezanzeb** [reply]
> we can probably keep the entire exception name in ctx is that would help.

Yes, please. Would it maybe be even possible to have a reference to the error class in there instead of a string? Or a reference to the original exception? Because if two files define an exception class with the same name, the name would be ambiguous.
